node/uhpo/src/payout_update.rs

```diff
@@ -0,0 +1,275 @@
+            &coinbase_txout,
+            prev_update_txout.as_ref(),
+        )?;
+
+        Ok(PayoutUpdate {
+            transaction: payout_update_tx,
+            coinbase_txout,
+            prev_update_txout,
+        })
+    }
+
+    pub fn add_coinbase_sig(
+        &mut self,
+        private_key: &SecretKey,
+        tweak: &Option<Scalar>,
+    ) -> Result<(), UhpoError> {
+        let prevouts = match &self.prev_update_txout {
+            Some(prev_update_txout) => vec![&self.coinbase_txout, prev_update_txout],
+            None => vec![&self.coinbase_txout],
+        };
+
+        add_signature(&mut self.transaction, 0, &prevouts, private_key, tweak)
+    }
+
+    pub fn add_prev_update_sig(
+        &mut self,
+        private_key: &SecretKey,
+        tweak: &Option<Scalar>,
+    ) -> Result<(), UhpoError> {
+        let prev_update_txout = self
+            .prev_update_txout
+            .as_ref()
+            .ok_or(UhpoError::NoPrevUpdateTxOut)?;
+        let prevouts = vec![&self.coinbase_txout, prev_update_txout];
+
+        add_signature(&mut self.transaction, 1, &prevouts, private_key, tweak)
+    }
+
+    pub fn build(self) -> Transaction {
+        self.transaction
+    }
+}
+
+fn build_transaction(
+    coinbase_tx: &Transaction,
+    prev_update_tx: Option<&Transaction>,
+    next_out_address: Address,
+    projected_fee: Amount,
+    coinbase_txout: &TxOut,
+    prev_update_txout: Option<&TxOut>,
+) -> Result<Transaction, Error> {
+    let mut total_amount = coinbase_txout.value;
+    if let Some(tx_out) = prev_update_txout {
+        total_amount += tx_out.value;
+    }
+
+    let mut payout_update_tx = Transaction {
+        version: Version::TWO,
+        lock_time: LockTime::ZERO,
+        input: vec![],
+        output: vec![],
+    };
+
+    payout_update_tx.input.push(TxIn {
+        previous_output: OutPoint {
+            txid: coinbase_tx.compute_txid(),
+            vout: 0,
+        },
+        script_sig: ScriptBuf::new(),
+        sequence: Sequence::MAX,
+        witness: Witness::default(),
+    });
+
+    if let Some(tx) = prev_update_tx {
+        payout_update_tx.input.push(TxIn {
+            previous_output: OutPoint {
+                txid: tx.compute_txid(),
+                vout: 0,
+            },
+            script_sig: ScriptBuf::new(),
+            sequence: Sequence::MAX,
+            witness: Witness::default(),
+        });
+    }
+
+    payout_update_tx.output.push(TxOut {
+        value: total_amount - projected_fee,
+        script_pubkey: next_out_address.script_pubkey(),
+    });
+
+    Ok(payout_update_tx)
+}
+
+fn add_signature(
+    transaction: &mut Transaction,
+    input_idx: usize,
+    prevouts: &[&TxOut],
+    private_key: &SecretKey,
+    tweak: &Option<Scalar>,
+) -> Result<(), UhpoError> {
+    let secp = Secp256k1::new();
+
+    let keypair: Keypair = match tweak {
+        Some(tweak) => private_key.keypair(&secp).add_xonly_tweak(&secp, tweak)?,
```

Comment:
We should differentiate the error here and the one on line 158, so that the client knows where the signing failed.

Reply:
Added new errors as suggested

- UhpoError::KeypairCreationError
- UhpoError::SignatureVerificationError